File: prob2.java
// Time Complexity : O(n)
// Space Complexity : O(1)

import java.util.*;

class GFG {

    public static void main(String[] args) {
        List<List<Double>> levels = new ArrayList<>();
        levels.add(Arrays.asList(10000.0, 0.3));
        levels.add(Arrays.asList(20000.0, 0.2));
        levels.add(Arrays.asList(30000.0, 0.1));
        levels.add(Arrays.asList(null, 0.1));
        
        double tax = calculateTax(levels, 45000);
        System.out.println(tax);
    }

    public static double calculateTax(List<List<Double>> levels, double salary) {
        double tax = 0;
        double previousLevel = 0;
        
        for (List<Double> level : levels) {
            if (level.get(0) == null || salary <= level.get(0)) {
                tax += (salary - previousLevel) * level.get(1);
                break;
            } else {
                tax += (level.get(0) - previousLevel) * level.get(1);
                previousLevel = level.get(0);
            }
        }
        
        return tax;
    }
}

File: prob2.py
# Time Complexity : O(n)
# Space Complexity :O(1)
# Passed on Leetcode: yes

class Solution:
    def calculateTax(self, brackets: List[List[int]], income: int) -> float:
        tax = 0

        for i in range(len(brackets)):
            uL = brackets[i][0]
            if  i == 0:
                if income >= uL:
                    taxable = uL
                    tax = tax + (taxable * (brackets[i][1]/100))
                else:
                    taxable = income
                    tax = tax + (taxable * (brackets[i][1]/100))
                    return tax
            else:
                if income >= uL:
                    taxable = brackets[i][0] - brackets[i-1][0]
                    tax = tax + (taxable * (brackets[i][1]/100))
                else:
                    taxable = income - brackets[i-1][0]
                    tax = tax + (taxable * (brackets[i][1]/100))
                    return tax
                    
        return tax